manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,740 @@
+the \emph{execution trace}. There might be several execution traces for a given
+program due to non-determinism.
+
+For a given execution trace, we define an irreflexive, transitive
+\emph{happens-before relation} that captures the causality between actions in
+the OCaml program. The happens-before relation is defined as follows:
+
+\begin{itemize}
+  \item We define the order in which a domain executes its actions as the
+    \emph{program order}. If an action "x" precedes another action "y" in
+    program order, then "x" precedes "y" in happens-before order.
+  \item If "x" is an initialising write to a memory location (atomic or
+    otherwise) and "y" is any read or write to that memory location, then "x"
+    precedes "y" in happens-before order.
+  \item If "x" is a write to an atomic location and "y" is a subsequent read or
+    write to that memory location in the execution trace, then "x" precedes "y"
+    in happens-before order. For atomic locations, "compare_and_set",
+    "fetch_and_add", "exchange", "incr" and "decr" are considered to perform
+    both a read and a write.
+  \item If "x" is "Domain.spawn f" and "y" is the first action in the newly
+    spawned domain executing "f", then "x" precedes "y" in happens-before
+    order.
+  \item If "x" is the last action in a domain "d" and "y" is "Domain.join
+    d", then "x" precedes "y" in happens-before order.
+  \item If "x" is an unlock operation on a mutex, and "y" is any subsequent
+    operation on the mutex in the execution trace, then "x" precedes "y" in
+    happens-before order.
+\end{itemize}
+
+\subsection{s:datarace}{Data race}
+
+In a given trace, two actions are said to be \emph{conflicting} if they access
+the same non-atomic location, and at least one is a write.
+
+We say that a program has a \emph{data race} if there exists some execution
+trace of the program with two conflicting actions and there does not exist a
+happens-before relationship between the conflicting accesses. A program without
+data races is said to be \emph{correctly synchronised}.
+
+\subsection{ss:drf_sc}{DRF-SC}
+
+\textbf{DRF-SC guarantee:} A program without data races will only exhibit
+sequentially consistent behaviours.
+
+DRF-SC is a strong guarantee for the programmers. Programmers can use
+sequential reasoning (executing one inter-domain action to identify whether
+their program has a data race. In particular, they do not need to reason about
+reorderings described in section~\ref{s:why_relaxed_memory} in order to determine
+whether their program has a data race. Once the determination that a particular
+program is data race free is made, they do not need to worry about reorderings
+in their code.
+
+For programs with data races, OCaml memory model ensures that there are no
+out-of-thin-air (OOTA) values. Reads always return a value written to a memory
+location (including possibly be the initial write). Data races do not cause
```

Comment:
```suggestion
out-of-thin-air (OOTA) values. Reads always return a value written to that memory
location. Data races do not cause
```
(I assume "including possibly be the initial write" means "an initial write is a write", but that seems obvious from the name)

Reply:
Notice that even in the archetypal example of OOTA, every read returns a value written by some write:
```
Initially *x=0, *y=0

r0 = *x  | r1 = *y
*y = r0  | *x = r1
exists r0 == 42 && r1 == 42
```
For instance the read  `r0 = *x` reads from write `*x = r1` .

Anyway. defining OOTA is tricky (is there a definition beyond examples?).

Reply:
We don't have a definition of OOTA. Given that the OCaml memory model does not exhibit OOTA, it may be better to drop the reference to OOTA from the manual.